### PR TITLE
Add pestpp

### PR DESCRIPTION
**Pull request by mwtoews**
Checklist
- [X] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [X] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [X] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [X] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [X] Build number is 0.
- [X] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by mwtoews** [comment]
@conda-forge/help-c-cpp, ready for review!

**Comment by mwtoews** [comment]
### For reviewers

- The source package includes a few vendored components:
  - `src/libs/Eigen`: aka https://anaconda.org/conda-forge/eigen / https://github.com/conda-forge/eigen-feedstock via [MPL2](http://opensource.org/search/node/MPL2) licence. Would it be preferred to use the library from Conda instead?
  - `src/libs/opt`: this directory is a combination of a few components from https://www.coin-or.org/, specifically:
    -  https://anaconda.org/conda-forge/coin-or-clp / https://github.com/conda-forge/coin-or-clp-feedstock via [EPL-2.0](http://opensource.org/search/node/EPL-2.0), and
    - https://anaconda.org/conda-forge/coin-or-utils / https://github.com/conda-forge/coin-or-utils-feedstock via [EPL-2.0](http://opensource.org/search/node/EPL-2.0)
    - There is no issue compiling this on mac/win/linux, however the conda-forge pacakges are not available for win (not sure why)
  - Question for the vendored packages: how to specify the licences for these packages in `meta.yaml`? The documented examples don't show this.
- The test section does not do much. Unfortunately, there is no simple `--version` flag to test a simple output from the programs. The programs expect to run a heavy simulation, requiring many other programs to operate. More exhaustive testing will need to be done after.
- Further info: I'm not a primary author for this package, but I wrote the CMake setup for the package a few years ago.

**Comment by carterbox** [comment]
>   * `src/libs/Eigen`: aka https://anaconda.org/conda-forge/eigen / https://github.com/conda-forge/eigen-feedstock via [MPL2](http://opensource.org/search/node/MPL2) licence. Would it be preferred to use the library from Conda instead?

Eigen is one of the special libraries that is header only. Header only libraries, if you don't use the conda package, you need to add their license to the license section of this recipe.

>   * `src/libs/opt`: this directory is a combination of a few components from https://www.coin-or.org/, specifically:
>     
>     * https://anaconda.org/conda-forge/coin-or-clp / https://github.com/conda-forge/coin-or-clp-feedstock via [EPL-2.0](http://opensource.org/search/node/EPL-2.0), and
>     * https://anaconda.org/conda-forge/coin-or-utils / https://github.com/conda-forge/coin-or-utils-feedstock via [EPL-2.0](http://opensource.org/search/node/EPL-2.0)
>     * There is no issue compiling this on mac/win/linux, however the conda-forge pacakges are not available for win (not sure why)

Libraries that are not header-only should be linked dynamically and use conda packages. If you need a Windows build for that existing package you should work with that existing feedstock to enable a Windows build. 

>   * Question for the vendored packages: how to specify the licences for these packages in `meta.yaml`? The documented examples don't show this.

List the path of each file.

```
license: CC0-1.0 AND [revendored licenses]
license_file:
  - USGS_disclaimer.txt
  - src/libs/Eigen/COPYING.BSD
  - src/libs/Eigen/COPYING.GPL
  - ....
```

The source tarball is unpacked before conda-build searches for license files.


> * The test section does not do much. Unfortunately, there is no simple `--version` flag to test a simple output from the programs. The programs expect to run a heavy simulation, requiring many other programs to operate. More exhaustive testing will need to be done after.

Unit and Correctness testing is the responsibility of upstream. The tests in this recipe are for *packaging*. They just need to check that the library was compiled, linked, installed correctly. Sometimes existence testing is sufficient.

https://conda-forge.org/docs/maintainer/adding_pkgs.html#test

File: recipes/pestpp/build.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+mkdir -p build && cd build
+
+cmake ${CMAKE_ARGS} \
+      -D CMAKE_BUILD_TYPE=Release \
+      -D BUILD_SHARED_LIBS=OFF \
```
Comment (carterbox) [inline recipes/pestpp/build.sh:7]:
```suggestion
      -D BUILD_SHARED_LIBS=ON \
```
See explanation elsewhere about static linking.

File: recipes/pestpp/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    - {{ compiler('cxx') }}
+    - cmake
+    - make   # [not win]
+    - ninja  # [win]
+
+test:
+  commands:
+    - ldd $(which pestpp-ies)  # [linux]
+    - otool -L $(which pestpp-ies)  # [osx]
+    - where pestpp-ies  # [win]
```
Comment (carterbox) [inline recipes/pestpp/meta.yaml:25]:
```suggestion
  commands:
    - ldd $(which pestpp-ies)  # [linux]
    - otool -L $(which pestpp-ies)  # [osx]
    - where pestpp-ies  # [win]
```
Typically, we use existence testing to check that libs, headers, binaries, and CMake artifacts are installed where expected. Do that instead.

**Review by carterbox — CHANGES_REQUESTED** [review]

**Comment by mwtoews** [comment]
Thanks @carterbox for your detailed review! We're aiming to address several of your points raised, but they may take a while to materialize.

**Pull request closed without merging**